File: src/features/returns.py
"""Returns and trend features (momentum backbone)."""

from typing import Any

import numpy as np
import pandas as pd

from .base import (
    EPS,
    BaseFeatureCalculator,
    FeatureSpec,
    ema,
    log_return,
    rolling_regression_slope,
    safe_divide,
)


class ReturnFeatureCalculator(BaseFeatureCalculator):
    """Computes return and trend features.

    Features:
        - r1: 1-bar log return
        - r5: 5-bar log return
        - r15: 15-bar log return
        - r60: 60-bar log return
        - cumret_60: Cumulative return over 60 bars
        - ema_diff: Normalized EMA difference (EMA12 - EMA48) / price
        - slope_60: Linear regression slope of log prices
        - trend_strength: |slope_60| / rv_60 (trend vs volatility)

    These features capture direction and persistence at multiple horizons.
    """

    def __init__(
        self,
        short_window: int = 5,
        medium_window: int = 15,
        long_window: int = 60,
        ema_fast: int = 12,
        ema_slow: int = 48,
    ):
        """Initialize ReturnFeatureCalculator.

        Args:
            short_window: Short-term return window (default 5)
            medium_window: Medium-term return window (default 15)
            long_window: Long-term window for slope, cumret, etc. (default 60)
            ema_fast: Fast EMA period (default 12)
            ema_slow: Slow EMA period (default 48)
        """
        self.short_window = short_window
        self.medium_window = medium_window
        self.long_window = long_window
        self.ema_fast = ema_fast
        self.ema_slow = ema_slow

    @property
    def feature_specs(self) -> list[FeatureSpec]:
        return [
            FeatureSpec(
                name="r1",
                description="1-bar log return",
                lookback=2,
                group="returns",
            ),
            FeatureSpec(
                name="r5",
                description="5-bar log return",
                lookback=self.short_window + 1,
                group="returns",
            ),
            FeatureSpec(
                name="r15",
                description="15-bar log return",
                lookback=self.medium_window + 1,
                group="returns",
            ),
            FeatureSpec(
                name="r60",
                description="60-bar log return",
                lookback=self.long_window + 1,
                group="returns",
            ),
            FeatureSpec(
                name="cumret_60",
                description="Cumulative log return over 60 bars",
                lookback=self.long_window,
                group="returns",
            ),
            FeatureSpec(
                name="ema_diff",
                description="Normalized EMA difference: (EMA12 - EMA48) / price",
                lookback=self.ema_slow,
                group="returns",
            ),
            FeatureSpec(
                name="slope_60",
                description="Linear regression slope of log prices over 60 bars",
                lookback=self.long_window,
                group="returns",
            ),
            FeatureSpec(
                name="trend_strength",
                description="Trend strength: |slope_60| / rv_60",
                lookback=self.long_window,
                group="returns",
            ),
        ]

    def compute(self, df: pd.DataFrame, **kwargs: Any) -> pd.DataFrame:
        """Compute return and trend features.

        Args:
            df: DataFrame with column: close
            **kwargs:
                rv_60: Optional pre-computed 60-bar realized volatility.
                       If not provided, will be computed from r1.

        Returns:
            DataFrame with columns: r1, r5, r15, r60, cumret_60, ema_diff,
                                   slope_60, trend_strength
        """
        close = df["close"]

        result = pd.DataFrame(index=df.index)

        # Log returns at various horizons
        result["r1"] = log_return(close, periods=1)
        result["r5"] = log_return(close, periods=self.short_window)
        result["r15"] = log_return(close, periods=self.medium_window)
        result["r60"] = log_return(close, periods=self.long_window)

        # Cumulative return over long window (sum of r1)
        result["cumret_60"] = result["r1"].rolling(
            window=self.long_window, min_periods=self.long_window
        ).sum()

        # EMA difference (trend proxy)
        ema_fast = ema(close, span=self.ema_fast)
        ema_slow = ema(close, span=self.ema_slow)
        result["ema_diff"] = safe_divide(ema_fast - ema_slow, close)

        # Slope of log prices
        log_prices = np.log(close)
        result["slope_60"] = rolling_regression_slope(log_prices, window=self.long_window)

        # Trend strength: |slope| / volatility
        # Use rv_60 from kwargs if provided, otherwise compute from r1
        rv_60 = kwargs.get("rv_60")
        if rv_60 is None:
            rv_60 = result["r1"].rolling(
                window=self.long_window, min_periods=self.long_window
            ).std()

        result["trend_strength"] = safe_divide(result["slope_60"].abs(), rv_60)

        return result

File: indicator-engine/src/indicators/returns.cpp
#include "indicators/returns.h"
#include <cmath>
#include <spdlog/spdlog.h>

namespace ie {

void ReturnCalculator::compute(
    const std::vector<OHLCVBar>& bars,
    std::vector<IndicatorResult>& results,
    std::vector<double>& r1_out) const
{
    const int n = static_cast<int>(bars.size());
    r1_out.resize(n, NaN);

    EmaState ema_f, ema_s;
    ema_f.init(ema_fast);
    ema_s.init(ema_slow);

    RollingSum cumret_sum;
    cumret_sum.init(long_window);

    RollingSlope slope;
    slope.init(long_window);

    RollingStats rv_rolling;
    rv_rolling.init(long_window);

    for (int i = 0; i < n; i++) {
        auto& r = results[i].features;
        double c = bars[i].close;

        // r1: 1-bar log return
        double r1 = (i >= 1) ? log_return(c, bars[i - 1].close) : NaN;
        r["r1"] = r1;
        r1_out[i] = r1;

        // r5, r15, r60: multi-period log returns
        r["r5"]  = (i >= short_window)  ? log_return(c, bars[i - short_window].close)  : NaN;
        r["r15"] = (i >= medium_window) ? log_return(c, bars[i - medium_window].close) : NaN;
        r["r60"] = (i >= long_window)   ? log_return(c, bars[i - long_window].close)   : NaN;

        // cumret_60: rolling sum of r1 over 60 bars
        if (is_valid(r1)) cumret_sum.push(r1);
        r["cumret_60"] = cumret_sum.full() ? cumret_sum.sum() : NaN;

        // ema_diff: (EMA_fast - EMA_slow) / close
        double ef = ema_f.update(c);
        double es = ema_s.update(c);
        if (i >= ema_slow - 1 && c > 0.0) {
            r["ema_diff"] = safe_divide(ef - es, c);
        } else {
            r["ema_diff"] = NaN;
        }

        // slope_60: linear regression slope of log(close)
        double lc = (c > 0.0) ? std::log(c) : NaN;
        if (is_valid(lc)) slope.push(lc);
        r["slope_60"] = slope.full() ? slope.slope() : NaN;

        // trend_strength: |slope_60| / rv_60
        if (is_valid(r1)) rv_rolling.push(r1);
        double rv60 = rv_rolling.full() ? rv_rolling.std_dev() : NaN;
        if (is_valid(r["slope_60"]) && is_valid(rv60)) {
            r["trend_strength"] = safe_divide(std::abs(r["slope_60"]), rv60);
        } else {
            r["trend_strength"] = NaN;
        }
    }

    spdlog::debug("ReturnCalculator: computed {} bars", n);
}

} // namespace ie
